experimental/parser/lexer.go

```diff
@@ -0,0 +1,675 @@
+		if err == nil {
+			token.SetValue(tok, value)
+			return tok
+		}
+
+		if err.(*strconv.NumError).Err == strconv.ErrRange {
+			l.Error(ErrIntegerOverflow{Token: tok})
+		} else {
+			l.Error(ErrInvalidNumber{Token: tok})
+		}
+	}
+
+	// This is either a float or a decimal number. Try parsing as decimal first, otherwise
+	// try parsing as float.
+	_, err := strconv.ParseUint(digits, 10, 64)
+	if err == nil {
+		// This is the most common result. It gets computed ON DEMAND by token.Token.AsNumber.
+		// DO NOT place it into l.literals.
+		return tok
+	}
+
+	// Check if it's out of range, because if that's the only problem, it's
+	// guaranteed to parse as a possibly-infinite float.
+	if err.(*strconv.NumError).Err == strconv.ErrRange {
+		// We want this to overflow to Infinity as needed, which ParseFloat
+		// will do for us. Otherwise it will ties-to-even as expected. Currently,
+		// the spec does not say "ties-to-even", but it says "nearest value",
+		// which everyone says when they mean "ties-to-even" whether they know it
+		// or not.
+		//
+		// ParseFloat itself says it "returns the nearest floating-point number
+		// rounded using IEEE754 unbiased rounding", which is just a weird way to
+		// say "ties-to-even".
+		value, _ := strconv.ParseFloat(digits, 64)
+		token.SetValue(tok, value)
+		return tok
+	}
+
+	// If it was not well-formed, it might be a float.
+	value, err := strconv.ParseFloat(digits, 64)
+
+	// This time, the syntax might be invalid. If it is, we decide whether
+	// this is a bad float or a bad integer based on whether it contains
+	// periods.
+	if err.(*strconv.NumError).Err == strconv.ErrSyntax {
+		l.Error(ErrInvalidNumber{Token: tok})
+	}
+
+	// If the error is ErrRange we don't care, it will clamp to Infinity in
+	// the way we want, as noted by the comment above.
+	token.SetValue(tok, value)
+	return tok
+}
+
+// LexString lexes a string starting at the current cursor.
+//
+// The cursor position should be just before the string's first quote character.
+func (l *lexer) LexString() token.Token {
+	start := l.cursor
+	q := l.Pop()
+
+	// Seek to the end of the string, unescaping as we go. We do not
+	// materialize an unescaped string if this string does not require escaping.
+	var buf strings.Builder
+	var haveEsc bool
+escapeLoop:
+	for !l.Done() {
+		r := l.Pop()
+		if r == q {
+			break
+		}
+
+		// Warn if the user has a non-printable character in their string that isn't
+		// ASCII whitespace.
+		if !unicode.IsGraphic(r) && !strings.ContainsRune(" \n\t\r", r) {
+			l.Warnf("non-printable character in string literal").With(
+				report.Snippetf(l.SpanFrom(l.cursor-utf8.RuneLen(r)), "this is the rune U+%04x", r),
+			)
+		}
+
+		if r != '\\' {
+			// We intentionally do not legalize against literal \0 and \n. The above warning
+			// covers \0 and legalizing against \n is user-hostile. This is valuable for
+			// e.g. strings that contain CEL code.
```

Comment:
Similar to other remarks above: this would be a language change, and we should discuss that separately. Let's first implement the language as it is today, and then we can figure out later where we have consensus for language changes.

Like above, probably fine to accept string literals with newlines and nil characters (to enable any subsequent processing and diagnostics), but we should emit a diagnostic that the langauge  doesn't allow it.